README.md
```bash
                      __                                         __        
   ____   ____   ____/ /  ___          _____   __  __    ____   / /_  ____ 
  / __ \ / __ \ / __  /  / _ \ ______ / ___/  / / / /   / __ \ / __/ / __ \
 / / / // /_/ // /_/ /  /  __//_____// /__   / /_/ /   / /_/ // /_  / /_/ /
/_/ /_/ \____/ \__,_/   \___/        \___/   \__, /   / .___/ \__/  \____/ 
                                            /____/   /_/                   
```
[![Build Status](https://travis-ci.org/SensitiveMix/node-cypto.svg?branch=master)](https://travis-ci.org/SensitiveMix/node-cypto)
[![codecov](https://codecov.io/gh/SensitiveMix/node-cypto/branch/master/graph/badge.svg)](https://codecov.io/gh/SensitiveMix/node-cypto)



A module for taking advantage of the built-in node signature module in node v0.8 and should contain host private and public ssh key.

## Installation

[![Greenkeeper badge](https://badges.greenkeeper.io/SensitiveMix/node-cypto.svg)](https://greenkeeper.io/)
```bash
yarn install https://github.com/SensitiveMix/node-cypto
```


## Usage
Initialize Request Handler for Http Signature

```bash
const Crpto = require('phoenix-crpto').client
let cypto = new Crpto('privateKeys','host','port')
cypto
  .then((result)=>{
  console.log(`success return: ${result}`)
  })
  .catch(e=>{
  console.error(`fail return: ${e}`)
  })
```

## Test
```bash
yarn test
```



## License 

MIT License

Copyright (c) 2016 Jack Sun

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.




CHANGELOG.md
<a name="1.0.0"></a>
# 1.0.0 (2017-09-17)





test/request.test.js
const Request = require('../libs/client')
const should = require('should')
const expect = require('expect')
const privateKeys = require('./certificate/private.js')
const host = '127.0.0.1'
const port = '3001'
const nock = require('nock')(`http://${host}:${port}`)


describe('request signature unit test', function () {
    before(() => {
        nock
            .get('/verify')
            .reply(200, {"msg": "ok"})
    })

    it('should resolve request if valid header', function (done) {
        let request = new Request(privateKeys, host, port)
        request.sendReq('phoenix-test', 'GET', {key: '123'}, "/verify")
            .then((res) => {
                expect(res.statusCode).toBe(200)
                expect(res.body.msg).toBe('ok')
            })
        done()
    })
})

test/certificate/public.js
module.exports = {
    "phoenix-test": "./test/certificate/public.pem"
}

test/certificate/private.js
module.exports = {
    "phoenix-test": "./test/certificate/private.pem"
}

libs/server.js
const fs = require('fs')
const md5 = require('js-md5')
const ALLOWS_LAG_SECONDS = 30
const moment = require('moment')
const httpSignature = require('http-signature')

class HttpServerHandler {
    constructor(opt) {
        this.req = opt.req
        this.res = opt.res
        this.next = opt.next
        this.publicKeys = opt.publicKeys
    }

    sendRes() {
        if (!this.req.headers.date) {
            return res.status(400)
                .json({msg: 'date header is required'})
        }
        this._processRes([this._checkSignature(req), HttpServerHandler._checkHeaderDate(this.req.headers.date), this._checkContentSignature(req)])
    }

    _checkSignature() {
        let _parsed = httpSignature.parseRequest(this.req)
        let pub = this._loadPublicKey()
        return httpSignature.verifySignature(_parsed, pub)
    }

    _loadPublicKey() {
        this._publicKeys = {}
        for (let name in this.publicKeys) {
            this._publicKeys[name] = fs.readFileSync(this.publicKeys[name], 'ascii')
            if (!this._publicKeys[name]) {
                throw new Error('Load public.pem key failed.')
            }
        }
    }

    _checkContentSignature() {
        const contentMd5 = this.req.headers['content-md5']
        let content = '' + this.req.method.toLowerCase() + this.req._parsedUrl.path
        if (this.req.body) {
            content = content + JSON.stringify(this.req.body)
        }
        return md5(content) === contentMd5
    }

    static _checkHeaderDate(datetime) {
        let status = true
        if (!this.req.headers.date) {
            return res.status(400)
                .json({msg: 'date header is required'})
        }
        return moment(Date.parse(datetime)).add(ALLOWS_LAG_SECONDS, 'seconds').isAfter(moment(new Date()))
    }

    _processRes(..._args) {
        new Promise((resolve, reject) => {
            for (let point in _args) {
                if (!point) {
                    throw {status: 401, msg: 'authorization failed'}
                }
            }
            resolve(null,true)
        }).then((done)=>{this.res.send({})})
            .catch((err) => {
            this.res.status(err.status).json({msg: err.msg})
        })
    }
}

libs/requestHandler.js
const fs = require('fs')
const md5 = require('js-md5')
const httpSignature = require('http-signature')

class HttpSignatureRequest {
  constructor ({ privateKeys, host, port, useHttps = true, headers = {} }) {
    if (!privateKeys || Object.keys(privateKeys).length === 0) {
      throw new Error('privateKeys is required.')
    }
    if (!host) { throw new Error('`host` is required.') }

    this._loadPrivateKey(privateKeys)
    this.host = host
    this._useHttps = useHttps
    this.headers = headers
    if (port) { this._port = port }
  }

  get port () {
    if (this._port) { return this._port }
    if (this._useHttps) { return 443 }
    return 80
  }

  sendReq ({ keyId, method, postData, path, headers }) {
    if (keyId) {
      this._keyId = keyId
    } else {
      this._keyId = Object.keys(this.privateKeys)[0]
    }

    if (!method) { method = 'GET' }
    if (path) { this.path = path }
    if (headers) { this.headers = headers }

    if (!this.path) { throw new Error('`path` is required.') }

    let options = {
      method: method,
      host: this.host,
      path: this.path,
      port: this.port,
      headers: this.headers
    }

    const http = this._getHttp()
    this.headers['content-md5'] = this._getContentMd5(method, postData)
    this.headers['date'] = new Date(+new Date - 8 * 3600 * 1000)
    return new Promise((resolve, reject) => {
      this._req = http.request(options, (res) => { resolve(res) })
      this._signReq()
      if (postData && Object.keys(postData).length !== 0) {
        this._req.write(require('querystring').stringify(postData))
      }
      this._req.end()
    })
      .then(this._processRes)
  }

  _loadPrivateKey (privateKeys) {
    this._privateKeys = {}
    for (let name in privateKeys) {
      this._privateKeys[name] = fs.readFileSync(privateKeys[name], 'ascii')
      if (!this._privateKeys[name]) {
        throw new Error('Load private key failed.')
      }
    }
  }

  _getHttp () {
    return this._useHttps ? require('https') : require('http')
  }

  _getContentMd5 (method, postData) {
    let content = `${method.toLowerCase()}${this.path}`
    if (postData) {
      content = content + JSON.stringify(postData)
    }
    return md5(content)
  }

  _signReq () {
    httpSignature.sign(this._req, { key: this._privateKeys[this._keyId], keyId: this._keyId })
  }

  _processRes (res) {
    return new Promise((resolve, reject) => {
      let body = ''
      res.on('data', function (d) { body += d })
      res.on('end', function () {
        resolve({
          statusCode: res.statusCode,
          statusMessage: res.statusMessage,
          body: body
        })
      })
    })
  }
}


module.exports = HttpSignatureRequest

libs/client.js
const fs = require('fs')
const md5 = require('js-md5')
const httpSignature = require('http-signature')

class HttpSignatureRequest {
    constructor(privateKeys, host, port, useHttps = true, headers = {}) {
        if (!privateKeys || Object.keys(privateKeys).length === 0) {
            throw new Error('privateKeys is required.')
        }
        if (!host) {
            throw new Error('`host` is required.')
        }

        this._loadPrivateKey(privateKeys)
        this.host = host
        this._useHttps = useHttps
        this.headers = headers
        if (port) {
            this._port = port
        }
    }

    /**
     * Default port 80
     * @returns {*}
     */
    get port() {
        if (this._port) {
            return this._port
        }
        if (this._useHttps) {
            return 443
        }
        return 80
    }

    /**
     * handle request
     * @param keyId
     * @param method
     * @param postData
     * @param path
     * @param headers
     * @returns {Promise.<TResult>}
     */
    sendReq(keyId, method, postData, path, headers) {
        if (keyId) {
            this._keyId = keyId
        } else {
            this._keyId = Object.keys(this.privateKeys)[0]
        }


        if (!method) method = 'GET'
        if (path) this.path = path
        if (headers) this.headers = headers

        if (!this.path) throw new Error('`path` is required.')

        let options = {
            method: method,
            host: this.host,
            path: this.path,
            port: this.port,
            headers: this.headers
        }

        const http = this._getHttp()
        this.headers['content-md5'] = this._getContentMd5(method, postData)
        this.headers['date'] = new Date().toISOString()
        return new Promise((resolve, reject) => {
            this._req = http.request(options, (res) => {
                resolve(res)
            })
            this._signReq()
            if (postData && Object.keys(postData).length !== 0) {
                this._req.write(JSON.stringify(postData))
            }
            this._req.end()
        })
            .then(this._processRes)
    }

    /**
     * load private pem
     * @param privateKeys
     * @private
     */
    _loadPrivateKey(privateKeys) {
        this._privateKeys = {}
        for (let name in privateKeys) {
            this._privateKeys[name] = fs.readFileSync(privateKeys[name], 'ascii')
            if (!this._privateKeys[name]) {
                throw new Error('Load private.pem key failed.')
            }
        }
    }

    _getHttp() {
        return this._useHttps ? require('https') : require('http')
    }

    /**
     * md5 content
     * @param method
     * @param postData
     * @private
     */
    _getContentMd5(method, postData) {
        let content = `${method.toLowerCase()}${this.path}`
        if (postData) {
            content = content + JSON.stringify(postData)
        }
        return md5(content)
    }

    /**
     * signature
     * @private
     */
    _signReq() {
        httpSignature.sign(this._req, {key: this._privateKeys[this._keyId], keyId: this._keyId})
    }

    /**
     * parse server response
     * @param res
     * @returns {Promise}
     * @private
     */
    _processRes(res) {
        return new Promise((resolve, reject) => {
            let body = ''
            res.on('data', function (d) {
                body += d
            })
            res.on('end', function () {
                resolve({
                    statusCode: res.statusCode,
                    statusMessage: res.statusMessage,
                    body: JSON.parse(body)
                })
            })
        })
    }
}


module.exports = HttpSignatureRequest